src/main/java/com/coreos/jetcd/EtcdClient.java

```diff
@@ -1,28 +1,70 @@
 package com.coreos.jetcd;
 
+import com.coreos.jetcd.api.AuthGrpc;
+import com.coreos.jetcd.api.AuthenticateRequest;
+import com.coreos.jetcd.api.AuthenticateResponse;
 import com.coreos.jetcd.api.KVGrpc;
+import com.coreos.jetcd.exception.AuthFailedException;
+import com.coreos.jetcd.exception.ConnectException;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.protobuf.ByteString;
+import io.grpc.CallCredentials;
 import io.grpc.ManagedChannel;
 import io.grpc.ManagedChannelBuilder;
+import io.grpc.Metadata;
+import io.grpc.stub.MetadataUtils;
+
+import java.util.concurrent.ExecutionException;
 
 /**
  * Etcd Client
  */
 public class EtcdClient {
 
+    private static final String TOKEN = "token";
+
     private final ManagedChannelBuilder<?> channelBuilder;
     private final String[] endpoints;
     private final ManagedChannel channel;
 
     private final EtcdKV kvClient;
+    private final EtcdAuth authClient;
+
+    private KVGrpc.KVFutureStub kvStub;
+    private AuthGrpc.AuthFutureStub authStub;
+
+    public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder builder) throws ConnectException, AuthFailedException {
+        this.endpoints = new String[builder.endpoints().size()];
+        builder.endpoints().toArray(this.endpoints);
+        this.channelBuilder = channelBuilder != null ? channelBuilder : ManagedChannelBuilder.forAddress("localhost", 2379).usePlaintext(true);
 
-    public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder builder) {
-        this.endpoints = (String[])builder.endpoints().toArray();
-        this.channelBuilder = channelBuilder != null ? channelBuilder : ManagedChannelBuilder.forAddress("localhost", 2379);
         this.channel = this.channelBuilder.build();
 
-        KVGrpc.KVFutureStub kvStub = KVGrpc.newFutureStub(this.channel);
+        kvStub = KVGrpc.newFutureStub(this.channel);
+        authStub = AuthGrpc.newFutureStub(this.channel);
+
+        String token=null;
+
+        if(builder.getName()!=null && builder.getPassword()!=null){
+            token = authenticate(this.channel, builder.getName(), builder.getPassword());
+        }else if(!(builder.getName()==null&&builder.getPassword()==null))
+        {
+            throw new NullPointerException("Username or password can not be null.");
+        }
+        if(token!=null) {
+            /**
+             * add token to channel header
+             */
+            Metadata metadata = new Metadata();
+            Metadata.Key<String> TOKEN_KEY= Metadata.Key.of(TOKEN, Metadata.ASCII_STRING_MARSHALLER);
```

Comment:
This place has too many details and is confusing..
Can you help make them a function with a comprehensive function name?